File: python-engine/engine/normalizer.py
from __future__ import annotations

import re
from urllib.parse import urlparse

from ai.schemas import SearchPlan
from models.place import PlaceRecord


def normalize_phone(value: str | None) -> str | None:
    """Return international format: +923001234567"""
    if not value:
        return None
    s = re.sub(r"[^\d+]", "", value)
    if s.startswith("0092"):
        s = "+" + s[2:]
    elif s.startswith("92") and not s.startswith("+"):
        s = "+" + s
    elif s.startswith("03") and len(s) >= 10:
        s = "+92" + s[1:]
    elif s.startswith("0") and len(s) >= 10:
        # Landline: 042 1234567 -> +92421234567
        s = "+92" + s[1:]
    return s or None


def normalize_phone_national(value: str | None) -> str | None:
    """Return national format: 03001234567"""
    international = normalize_phone(value)
    if not international:
        return None
    # Strip the +92 prefix and prepend 0
    if international.startswith("+92"):
        return "0" + international[3:]
    return international


def normalize_website(value: str | None) -> str | None:
    if not value:
        return None
    value = value.strip()
    if not value:
        return None
    if not value.startswith(("http://", "https://")):
        value = "https://" + value
    return value


def normalize_text(value: str | None) -> str | None:
    if not value:
        return None
    return re.sub(r"\s+", " ", value).strip()


def normalize_place(raw: dict, plan: SearchPlan) -> PlaceRecord:
    display = raw.get("displayName") or {}
    location = raw.get("location") or {}
    types = raw.get("types") or []
    address = normalize_text(raw.get("formattedAddress"))
    area = None

    # Best-effort area extraction without inventing data.
    if address:
        parts = [p.strip() for p in address.split(",") if p.strip()]
        if len(parts) >= 3:
            area = parts[-3]

    raw_phone = raw.get("nationalPhoneNumber") or raw.get("internationalPhoneNumber")

    # Always extract coordinates (field mask now always includes places.location).
    lat = location.get("latitude") if location else None
    lng = location.get("longitude") if location else None

    return PlaceRecord(
        place_id=raw.get("id"),
        business_name=normalize_text(display.get("text")),
        category=plan.category,
        address=address,
        area=area,
        city=plan.location.city,
        district=plan.location.district,
        province=plan.location.province,
        country="Pakistan",
        phone=normalize_phone(raw_phone),
        phone_national=normalize_phone_national(raw_phone),
        website=normalize_website(raw.get("websiteUri")),
        google_maps_url=raw.get("googleMapsUri"),
        latitude=lat,
        longitude=lng,
        rating=raw.get("rating"),
        review_count=raw.get("userRatingCount"),
        business_status=raw.get("businessStatus"),
        source="Google Places API (New)",
        retrieved_at=raw.get("_retrieved_at"),
    )

File: lib/data-engine/normalizer.ts
import { SearchPlan, PlaceRecord } from "./types";

/**
 * Normalizes phone numbers to standard Pakistani international format: +923XXXXXXXXX
 */
export function normalizePhone(value: string | null | undefined): string | null {
  if (!value) return null;
  
  // Remove all non-digit and non-plus characters
  let s = value.replace(/[^\d+]/g, "");
  
  if (s.startsWith("0092")) {
    s = "+" + s.substring(2);
  } else if (s.startsWith("92") && !s.startsWith("+")) {
    s = "+" + s;
  } else if (s.startsWith("03") && s.length >= 10) {
    s = "+92" + s.substring(1);
  } else if (s.startsWith("0") && s.length >= 10) {
    // Landline: e.g. 0421234567 -> +92421234567
    s = "+92" + s.substring(1);
  }
  
  return s.startsWith("+") ? s : null;
}

/**
 * Normalizes phone numbers to standard Pakistani national format: 03XXXXXXXXX
 */
export function normalizePhoneNational(value: string | null | undefined): string | null {
  const international = normalizePhone(value);
  if (!international) return null;
  
  if (international.startsWith("+92")) {
    return "0" + international.substring(3);
  }
  return international;
}

/**
 * Normalizes website URLs to a standard comparison format.
 */
export function normalizeWebsite(value: string | null | undefined): string | null {
  if (!value) return null;
  const cleaned = value.trim();
  if (!cleaned) return null;
  
  if (!/^https?:\/\//i.test(cleaned)) {
    return "https://" + cleaned;
  }
  return cleaned;
}

/**
 * Normalizes a website to its base domain for deduplication.
 * e.g., "https://www.savourfoods.com.pk/location/" -> "savourfoods.com.pk"
 */
export function normalizeWebsiteDomain(url: string | null | undefined): string {
  if (!url) return "";
  try {
    let clean = url.trim().toLowerCase();
    // Remove protocol
    clean = clean.replace(/^(https?:\/\/)?(www\.)?/i, "");
    // Remove paths and query params
    clean = clean.split("/")[0].split("?")[0];
    return clean;
  } catch (e) {
    return "";
  }
}

/**
 * Normalizes text spacing.
 */
export function normalizeText(value: string | null | undefined): string | null {
  if (!value) return null;
  return value.replace(/\s+/g, " ").trim();
}

/**
 * Cleans a business name by removing common suffix abbreviations for fuzzy comparison.
 */
export function cleanBusinessName(name: string): string {
  return name
    .toLowerCase()
    .replace(/[.,\/#!$%\^&\*;:{}=\-_`~()]/g, "") // remove punctuation
    .replace(/\s+/g, " ") // normalize spaces
    .replace(/\b(pvt|ltd|limited|private|inc|co|company|group|restaurant|cafe|school|hotel|agency|realtors)\b/g, "")
    .trim();
}

/**
 * Map raw Places API New JSON fields to TypeScript PlaceRecord.
 */
export function normalizePlace(raw: any, plan: SearchPlan, retrievedAt?: string): PlaceRecord {
  const display = raw.displayName || {};
  const location = raw.location || {};
  const address = normalizeText(raw.formattedAddress);
  
  // Best effort area extraction
  let area: string | null = null;
  if (address) {
    const parts = address.split(",").map((p: string) => p.trim()).filter(Boolean);
    if (parts.length >= 3) {
      area = parts[parts.length - 3];
    }
  }

  const rawPhone = raw.nationalPhoneNumber || raw.internationalPhoneNumber;
  const lat = location.latitude ?? null;
  const lng = location.longitude ?? null;

  return {
    place_id: raw.id,
    business_name: normalizeText(display.text),
    category: plan.category,
    address,
    area,
    city: plan.location.city,
    district: plan.location.district,
    province: plan.location.province,
    country: "Pakistan",
    phone: normalizePhone(rawPhone),
    phone_national: normalizePhoneNational(rawPhone),
    website: normalizeWebsite(raw.websiteUri),
    google_maps_url: raw.googleMapsUri,
    latitude: lat,
    longitude: lng,
    rating: raw.rating ?? null,
    review_count: raw.userRatingCount ?? null,
    business_status: raw.businessStatus ?? null,
    source: "Google Places API (New)",
    retrieved_at: retrievedAt || new Date().toISOString()
  };
}
